### ISSUE TITLE
help understanding pod names

### ISSUE THREAD

**Issue by rabernat** [dask/dask-kubernetes#173]
In an attempt to control costs, am trying to figure out which users in my pangeo google cloud cluster have been creating dask kubernetes pods. All I have to work with are the kubernetes API calls logged by google cloud into bigquery.

The challenge is that it is hard to parse the names of the dask workers, which look something like this: `dask-jovyan-3569756c-e2j5ht`. I have concluded that the first random looking string is associated with the cluster, and the second is associated with the individual worker. Is there any hope of tracing this back to individual users?

**Comment by mrocklin** [comment]
`jovyan` *is* the username.  Dask-kubernetes looks at `$USER` and includes that within the job name for exactly your purpose.  Unfortunately, the system that you're running loads up docker images all with the same username.  You can configure this name with the following configuration entry:

https://github.com/dask/dask-kubernetes/blob/ae66baac4ba456dd58fac6e9070a385972bb3188/dask_kubernetes/kubernetes.yaml#L1-L2

Currently we use it like this:

https://github.com/dask/dask-kubernetes/blob/ae66baac4ba456dd58fac6e9070a385972bb3188/dask_kubernetes/core.py#L202-L204

**Comment by mrocklin** [comment]
So, if you have some way of getting the username (or some other identifier) into an environment variable, that would probably be your best bet.

**Comment by jhamman** [comment]
@rabernat - I think you can set the name to something like:

```yaml
      name: "dask-{JUPYTERHUB_USER}-{uuid}
```

Of course, this only works if each user follows instructions and set's their dask config correctly.

**Comment by rabernat** [comment]
Thanks for the suggestions.

Going forward, I think dask-gateway is the solution to these issues.

Retroactively, it sounds like there is no way to decode the login ID (as opposed to the unix username) from the logs.

**Comment by mrocklin** [comment]
I think that a broader solution would be to push on this upstream in JupyterHub.  

Ideally when users arrive in their session in JupyterHub their username reflects their login.

Ideally if you login to ocean.pangeo.io, you could do something like this

```
$ echo $USER
rabernat
```

Instead of the behavior today

```
$ echo $USER
jovyan
```

If this change was made, then everything else would follow naturally

**Comment by rabernat** [comment]
> - I think you can set the name to something like:
> 
> ```yaml
>       name: "dask-{JUPYTERHUB_USER}-{uuid}
> ```
> 
> Of course, this only works if each user follows instructions and set's their dask config correctly

@jhamman - what would be the best place to inject this into the config on ocean.pangeo.io?

We really need to find out who is spinning up 100-node dask clusters before we run out of money.

**Comment by jhamman** [comment]
@rabernat - that can be done here: https://github.com/pangeo-data/pangeo-stacks/blob/master/base-notebook/binder/dask_config.yaml#L21

Although, it actually seems like this is already setup. ocean.pangeo.io is using a fairly old base image so it may be worth upgrading since @scottyhq did some work on the dask configs since you last upgraded.

Some pointers:

1. The root config for images that use pangeo-stacks is at: `/srv/conda/envs/notebook/etc/dask/dask.yaml`
2. You can query the dask configuration headed to kubernetes via: 

```python
In [5]: from dask.config import expand_environment_variables                                                                       

In [6]: expand_environment_variables(dask.config.config['kubernetes'])                                                             
Out[6]: 
{'worker-template': {'metadata': None,
  'spec': {'restartPolicy': 'Never',
   'containers': [{'args': ['dask-worker',
      '--nthreads',
      '2',
      '--no-bokeh',
      '--memory-limit',
      '11.5GB',
      '--death-timeout',
      '60'],
     'image': 'us.gcr.io/pangeo-181919/ocean-pangeo-io-notebook:986ba85',
     'name': 'dask-0000-0001-7479-8439',
     'resources': {'limits': {'cpu': 16, 'memory': '11.5G'},
      'requests': {'cpu': 1, 'memory': '11.5G'}}}]}}}
```

**Comment by jacobtomlinson** [comment]
Going to close this as solutions seems to have been found.
